Registry login fails because wrong credentials are used

When trying to publish a container with a base image like `registry.git.human.de/docker/dotnet/aspnet:7.0` the wrong credentials will be used. It tries to use the credentials for `git.human.de` while it should be using `registry.git.human.de` instead.

The `~/.docker/config.json` looks like this:

```json
{
    "auths": {
        "registry.git.human.de": {
        "auth": "<secret>"
    }
}
```

# Reproduction

## Create a new web app

```bash
dotnet new webapp
```

## Add the library reference

```xml
<PackageReference Include="Microsoft.NET.Build.Containers" Version="0.2.7" />
```

## Configure the base image

```xml
<ContainerBaseImage>registry.git.human.de/docker/dotnet/aspnet:7.0</ContainerBaseImage>
```

## Try to publish the app

```bash
dotnet publish --os linux --arch x64 -c Release -p:PublishProfile=DefaultContainer
```

# Example output

```
MSBuild version 17.4.0+18d5aef85 for .NET
  Wiederherzustellende Projekte werden ermittelt...
  "/home/meenzens/dev/test/test.csproj" wiederhergestellt (in "152 ms").
  test -> /home/meenzens/dev/test/bin/Release/net7.0/linux-x64/test.dll
  test -> /home/meenzens/dev/test/bin/Release/net7.0/linux-x64/publish/
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018: Unerwarteter Fehler bei der CreateNewImage-Aufgabe. [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018: System.AggregateException: One or more errors occurred. (Failed retrieving credentials for "git.human.de": No matching auth specified for registry 'git.human.de' in Docker config '/home/meenzens/.docker/config.json'.) [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:  ---> Microsoft.NET.Build.Containers.Credentials.CredentialRetrievalException: Failed retrieving credentials for "git.human.de": No matching auth specified for registry 'git.human.de' in Docker config '/home/meenzens/.docker/config.json'. [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:  ---> Valleysoft.DockerCredsProvider.CredsNotFoundException: No matching auth specified for registry 'git.human.de' in Docker config '/home/meenzens/.docker/config.json'. [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    at Valleysoft.DockerCredsProvider.CredsProvider.GetCredStoreAsync(String registry, IFileSystem fileSystem, IProcessService processService, IEnvironment environment) [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    at Valleysoft.DockerCredsProvider.CredsProvider.GetCredentialsAsync(String registry, IFileSystem fileSystem, IProcessService processService, IEnvironment environment) [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    at Microsoft.NET.Build.Containers.AuthHandshakeMessageHandler.GetAuthenticationAsync(String scheme, Uri realm, String service, String scope, CancellationToken cancellationToken) in D:\a\_work\1\s\Microsoft.NET.Build.Containers\AuthHandshakeMessageHandler.cs:line 102 [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    --- End of inner exception stack trace --- [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    at Microsoft.NET.Build.Containers.AuthHandshakeMessageHandler.GetAuthenticationAsync(String scheme, Uri realm, String service, String scope, CancellationToken cancellationToken) in D:\a\_work\1\s\Microsoft.NET.Build.Containers\AuthHandshakeMessageHandler.cs:line 106 [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    at Microsoft.NET.Build.Containers.AuthHandshakeMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in D:\a\_work\1\s\Microsoft.NET.Build.Containers\AuthHandshakeMessageHandler.cs:line 167 [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken) [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    at Microsoft.NET.Build.Containers.Registry.GetImageManifest(String name, String reference) in D:\a\_work\1\s\Microsoft.NET.Build.Containers\Registry.cs:line 24 [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    --- End of inner exception stack trace --- [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions) [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification) [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    at Microsoft.NET.Build.Containers.Tasks.CreateNewImage.GetBaseImage() in D:\a\_work\1\s\Microsoft.NET.Build.Containers\CreateNewImage.cs:line 192 [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    at Microsoft.NET.Build.Containers.Tasks.CreateNewImage.Execute() in D:\a\_work\1\s\Microsoft.NET.Build.Containers\CreateNewImage.cs:line 208 [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/home/meenzens/dev/test/test.csproj]
/home/meenzens/.nuget/packages/microsoft.net.build.containers/0.2.7/build/Microsoft.NET.Build.Containers.targets(124,9): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/home/meenzens/dev/test/test.csproj]
```

--- a/Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs
+++ b/Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs
@@ -85,7 +85,7 @@ public partial class AuthHandshakeMessageHandler : DelegatingHandler
     /// <param name="scope"></param>
     /// <param name="cancellationToken"></param>
     /// <returns></returns>
-    private async Task<AuthenticationHeaderValue?> GetAuthenticationAsync(string scheme, Uri uri, string service, string? scope, CancellationToken cancellationToken)
+    private async Task<AuthenticationHeaderValue?> GetAuthenticationAsync(string registry, string scheme, Uri realm, string service, string? scope, CancellationToken cancellationToken)
     {
         // Allow overrides for auth via environment variables
         string? credU = Environment.GetEnvironmentVariable(ContainerHelpers.HostObjectUser);
@@ -102,18 +102,18 @@ public partial class AuthHandshakeMessageHandler : DelegatingHandler
         {
             try
             {
-                privateRepoCreds = await CredsProvider.GetCredentialsAsync(uri.Host);
+                privateRepoCreds = await CredsProvider.GetCredentialsAsync(registry);
             }
             catch (Exception e)
             {
-                throw new CredentialRetrievalException(uri.Host, e);
+                throw new CredentialRetrievalException(registry, e);
             }
         }
         
         if (scheme is "Basic")
         {
             var basicAuth = new AuthenticationHeaderValue("Basic", Convert.ToBase64String(Encoding.ASCII.GetBytes($"{privateRepoCreds.Username}:{privateRepoCreds.Password}")));
-            return AuthHeaderCache.AddOrUpdate(uri, basicAuth);
+            return AuthHeaderCache.AddOrUpdate(realm, basicAuth);
         }
         else if (scheme is "Bearer")
         {
@@ -121,7 +121,7 @@ public partial class AuthHandshakeMessageHandler : DelegatingHandler
             var header = privateRepoCreds.Username == "<token>"
                             ? new AuthenticationHeaderValue("Bearer", privateRepoCreds.Password)
                             : new AuthenticationHeaderValue("Basic", Convert.ToBase64String(Encoding.ASCII.GetBytes($"{privateRepoCreds.Username}:{privateRepoCreds.Password}")));
-            var builder = new UriBuilder(uri);
+            var builder = new UriBuilder(realm);
             var queryDict = System.Web.HttpUtility.ParseQueryString("");
             queryDict["service"] = service;
             if (scope is string s)
@@ -143,7 +143,7 @@ public partial class AuthHandshakeMessageHandler : DelegatingHandler
 
             // save the retrieved token in the cache
             var bearerAuth = new AuthenticationHeaderValue("Bearer", token.ResolvedToken);
-            return AuthHeaderCache.AddOrUpdate(uri, bearerAuth);
+            return AuthHeaderCache.AddOrUpdate(realm, bearerAuth);
         }
         else
         {
@@ -177,7 +177,7 @@ public partial class AuthHandshakeMessageHandler : DelegatingHandler
                 }
                 else if (response is { StatusCode: HttpStatusCode.Unauthorized } && TryParseAuthenticationInfo(response, out string? scheme, out AuthInfo? authInfo))
                 {
-                    if (await GetAuthenticationAsync(scheme, authInfo.Realm, authInfo.Service, authInfo.Scope, cancellationToken) is AuthenticationHeaderValue authentication)
+                    if (await GetAuthenticationAsync(request.RequestUri.Host, scheme, authInfo.Realm, authInfo.Service, authInfo.Scope, cancellationToken) is AuthenticationHeaderValue authentication)
                     {
                         request.Headers.Authorization = AuthHeaderCache.AddOrUpdate(request.RequestUri, authentication);
                         return await base.SendAsync(request, cancellationToken);

